### PR TITLE
Fix unit tests

### PR DESCRIPTION
**Pull request by jdkandersson**
- Refactor code for improved maintenance
- Get unittest coverage to 100%

### CODE REVIEW

**Comment by jdkandersson** [comment]
This is on top of the previous pull requests, once that is merged the changes from that PR will disappear from this PR.

File: src/charm.py
Comment (merkata) [inline src/charm.py:40]: Isn't it recommended to have your handlers declared as private i.e. starting with an underscore? https://juju.is/docs/sdk/styleguide#heading--event-handler-visibility

File: src/charm.py
Comment (merkata) [inline src/charm.py:54]: typo s/Tje/The

File: src/charm.py
Comment (merkata) [inline src/charm.py:76]: typo double "n"

**Comment by merkata** [comment]
Looks really good, solid work! I left just small remarks. 
There is one thing that I wanted to discuss - I see you run `tox -e unit`, and that is good - I was using it as well when converting a charm. I had to transfer some files (charmcraft.yaml most notably) to be able to `charmcraft pack` the converted charm. That's when I noticed the presence of `./run_tests` - it is generated by default for new charms by `charmcraft init`. It runs flake8 and unittest with coverage report and doesn't seem to rely on tox in any way. I'm wondering what is the recommended way of testing charms, as https://juju.is/docs/sdk/styleguide#heading--testing still favors tox.

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:40]: Thank you, fixed.

> Looks really good, solid work! I left just small remarks.
> There is one thing that I wanted to discuss - I see you run `tox -e unit`, and that is good - I was using it as well when converting a charm. I had to transfer some files (charmcraft.yaml most notably) to be able to `charmcraft pack` the converted charm. That's when I noticed the presence of `./run_tests` - it is generated by default for new charms by `charmcraft init`. It runs flake8 and unittest with coverage report and doesn't seem to rely on tox in any way. I'm wondering what is the recommended way of testing charms, as https://juju.is/docs/sdk/styleguide#heading--testing still favors tox.

Yeh I noticed that too, in this repo there is also a makefile which seems to have some value. My preference is also tox rather than the ./run_tests command since there is a range of tests (unit and integration) which are easily managed by tox. The indico charm is also using tox, maybe @mthaddon  can enlighten us :)

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: Thank you, fixed

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: Thank you, fixed

**Comment by jdkandersson** [comment]
> Looks really good, solid work! I left just small remarks.
> There is one thing that I wanted to discuss - I see you run `tox -e unit`, and that is good - I was using it as well when converting a charm. I had to transfer some files (charmcraft.yaml most notably) to be able to `charmcraft pack` the converted charm. That's when I noticed the presence of `./run_tests` - it is generated by default for new charms by `charmcraft init`. It runs flake8 and unittest with coverage report and doesn't seem to rely on tox in any way. I'm wondering what is the recommended way of testing charms, as https://juju.is/docs/sdk/styleguide#heading--testing still favors tox.

Yeh I noticed that too, in this repo there is also a makefile which seems to have some value. My preference is also tox rather than the ./run_tests command since there is a range of tests (unit and integration) which are easily managed by tox. The indico charm is also using tox, maybe @mthaddon  can enlighten us :)

**Comment by mthaddon** [comment]
> Yeh I noticed that too, in this repo there is also a makefile which seems to have some value. My preference is also tox rather than the ./run_tests command since there is a range of tests (unit and integration) which are easily managed by tox. The indico charm is also using tox, maybe @mthaddon can enlighten us :)

In the IS team we favoured having a `Makefile` that then does whatever's needed for a particular target (typically it just calls `tox` targets). At some time after this, the `charmcraft init` template added `./run_tests` which relies on sourcing a virtualenv first, which seems a little cumbersome to me. I think using `tox` is definitely the way to go here. We can remove the `Makefile` as well, and then just document how to develop/test locally on a documentation page for the charm similar to how we do on https://charmhub.io/indico/docs/contributing by updating https://charmhub.io/jenkins-agent-k8s/docs/contributing-hacking.

**Comment by merkata** [comment]
I asked about the `./run_tests` on starcraft channel and it seems that will go away and the preferred way is to use tox. Wrapping it in a Makefile like IS does is fine with me, I like Makefiles. Point here is that tox is preferred over ./run_tests and is aligned with the documented practice in the SDK https://juju.is/docs/sdk/styleguide#heading--testing

**Comment by jdkandersson** [comment]
The makefiles seem handy for things like building the images and then running tox with the right resources etc, reduces the mental overhead of having to remember how to do things. I would prefer keeping it, although it does require some cleanup for this repo (for which I'll add an issue).

@merkata / @mthaddon please complete the review, there is another PR for linting that is waiting on this one

File: src/charm.py
Comment (mthaddon) [inline src/charm.py:95]: We shouldn't use format strings in logging messages (I realise it was doing this before as well). See https://juju.is/docs/sdk/styleguide#heading--templating.

File: src/charm.py
Comment (mthaddon) [inline src/charm.py:162]: Same comment as above about logging and building strings.

File: src/charm.py
Comment (mthaddon) [inline src/charm.py:171]: Same comment as above about logging and building strings.

File: tests/unit/test_charm.py
Comment (mthaddon) [inline tests/unit/test_charm.py:10]: Same comment as above about explicit imports rather than importing "model" or "testing" for instance.

**Review by mthaddon — COMMENTED** [review]
Some comments inline

File: .github/workflows/ci.yaml
Comment (jdkandersson) [reply .github/workflows/ci.yaml:7]: Yes, the reason the Python installation action is used here is to guarantee the Python version being used, otherwise the workflow is going to use whatever is available in ubuntu-latest by default (which was 3.8 in this case which caused the tests to fail due to using some Python 3.10 features in this charm around type hinting. I'm happy to propose a pull request for the indigo charm to use the Python action there as well

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:9]: Yes this was before the I had included listing and formatting, the next PR that includes linting and fixes all these problems. I propose that we let this go through and fix it in the next PR (which is already ready just waiting for this one to go through).

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:11]: Importing the module helps make it clear where a class or function came from and clearly differentiates imported versus local functions. The Google style guide has some more thoughts on it: https://google.github.io/styleguide/pyguide.html#22-imports

The contribution guide has an issue on this topic, I propose we debate it there and then update our repos accordingly: https://github.com/canonical/is-charms-contributing-guide/issues/7

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:17]: Oops, thanks for pointing this out, I'll fix it

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:27]: Thanks I'll fix this

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:95]: Yes, I actually had fixed this in the next PR

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:162]: Yes, I actually had fixed this in the next PR

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:171]: Yes, I actually had fixed this in the next PR

File: tests/unit/test_charm.py
Comment (jdkandersson) [reply tests/unit/test_charm.py:10]: Let's discuss it there and I'll adjust whatever outcome we reach there here as well

File: .github/workflows/ci.yaml
Comment (mthaddon) [reply .github/workflows/ci.yaml:7]: The `charmcraft.yaml` has "build on" and "run on" 20.04. Will that cause any issues here if we're depending on 3.10 features?

Comment (jdkandersson) [reply .github/workflows/ci.yaml:7]: Yes at it turns out when I started the integration tests, I have since updated that to 22.04 and that'll be part of the PR for the integration tests

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: I have reverted this back

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: Fixed

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:164]: Fixed

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py:173]: Fixed

File: tests/unit/test_charm.py
Comment (jdkandersson) [reply tests/unit/test_charm.py]: Reverted it back

File: metadata.yaml
Comment (arturo-seijas) [inline metadata.yaml:7]: Should we add the team as maintainer?
```
  - launchpad.net/~canonical-is-devops
```

File: pyproject.toml
Comment (arturo-seijas) [inline pyproject.toml:8]: This is not aligned with the [template repository](https://github.com/canonical/template-operator). Do you think you could open a PR in there with the changes and discuss them if needed?

File: src/charm.py
Comment (arturo-seijas) [inline src/charm.py:99]: Won't `pebble.replan_services` achieve the same result without the need to check if there are changes?

**Review by arturo-seijas — CHANGES_REQUESTED** [review]

File: src/charm.py
Comment (jdkandersson) [reply src/charm.py]: Thank you, fixed

File: pyproject.toml
Comment (jdkandersson) [reply pyproject.toml:8]: I have updated it to be the same as the template project

File: metadata.yaml
Comment (jdkandersson) [reply metadata.yaml:8]: Thank you, done

**Review by arturo-seijas — APPROVED** [review]
LGTM

File: .github/workflows/ci.yaml
Comment (mthaddon) [inline .github/workflows/ci.yaml:15]: Does this not need to be 3.8 given our charmcraft.yaml uses 20.04?

**Review by mthaddon — COMMENTED** [review]
Looks great, just one minor question about python version in CI.

File: .github/workflows/ci.yaml
Comment (jdkandersson) [reply .github/workflows/ci.yaml:15]: Yeh I need to fix the `charmcraft.yaml` file to 22.04 and then also update the CI, thanks for the comment.

**Pull request merged**
